Create Search Bar to Search Posts

Leoism opened the pull request:
+ Allow users to search a query within all posts currently available
+ Return a list of all posts with the query in priority order based on: school name, then title, then text content
+ Add tests to ensure posts are retrieved in the expected manner

Commit:
Create Search Bar to Search Posts
+ Allow users to search a query within all posts currently available
+ Return a list of all posts with the query in priority order based on: school name, then title, then text content
+ Add tests to ensure posts are retrieved in the expected manner

andrewnguyen commented on capstone_project/src/main/java/com/google/sps/data/PostFilter.java:
```diff
@@ -0,0 +1,77 @@
+      }
+    }
+
+    postsFound.sort(PostFilter::compare);
```
This feels kind of odd. I think you should move the logic on lines 48-53 into `compare`. You can also consider using removeIf https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#removeIf-java.util.function.Predicate-

andrewnguyen replied:
basically... I don't like adding things to posts to just then removing them again in the same method.

lyndapham commented on capstone_project/src/main/java/com/google/sps/data/PostFilter.java:
```diff
@@ -0,0 +1,77 @@
+    for (Entity post : allPosts) {
+      String postSchool = ((String) post.getProperty("schoolName")).toLowerCase();
+      String postTitle = ((String) post.getProperty("title")).toLowerCase();
+      String postText = ((Text) post.getProperty("text")).getValue().toLowerCase();
```
Why do you cast to Text, then call getValue(), which seems to return a String? Can you cast directly to String?

Leoism replied:
No. When post.getProperty gets called it returns an Object class. the method getValue does not exist inside of the Object class. Therefore, I have to cast it into a Text object. From there, I can now call the getValue method which returns the entire contents of the Text object

lyndapham replied:
Thanks for the explanation!

andrewnguyen commented on capstone_project/src/main/webapp/scripts/search_bar.js:
```diff
@@ -0,0 +1,37 @@
+ */
+async function searchPosts() { // eslint-disable-line no-unused-vars
+  // Only run when the user presses enter.
+  if (event.keyCode !== 13) {
```
Can you make a reference to the standard keycode string? https://w3c.github.io/uievents-code/#key-alphanumeric-functional

rtnavarr commented on capstone_project/src/main/java/com/google/sps/data/PostFilter.java:
```diff
@@ -0,0 +1,77 @@
+public class PostFilter {
+  public static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  private PostFilter() {}
```
is having an empty constructor necessary?

Leoism replied:
Since all methods in this class are static, a private constructor prevents the instantiation of the class directly. Having it empty and private means it can't be called outside the class and nothing will happen if it is called.

lyndapham commented on capstone_project/src/test/java/com/google/sps/PostFilterTest.java:
```diff
@@ -0,0 +1,123 @@
+    List<Entity> postEntities = Arrays.asList(post3, post2, post1);
+    List<Post> expectedPostOrder =
+        postEntities.stream().map(PostService::convertEntityToPost).collect(Collectors.toList());
+    String expectedGson = new Gson().toJson(expectedPostOrder);
```
Would this technically be more aptly named as expectedJson?

lyndapham commented on capstone_project/src/test/java/com/google/sps/PostFilterTest.java:
```diff
@@ -0,0 +1,123 @@
+
+  private final String SEARCH_QUERY = "Test University";
+  private final String DEFAULT_STRING = "Some Text";
+  private Entity post1;
```
Not necessary, but it might be helpful to comment either here, or in the set up, a very short description about each test post, or to briefly explain the expected values in comments before each test. Reading the set up is somewhat difficult just because there is a lot of text.

Leoism replied:
Done

lyndapham commented on capstone_project/src/test/java/com/google/sps/PostFilterTest.java:
```diff
@@ -0,0 +1,123 @@
+  @Test
+  public void returnsEmptyArrayOnNoMatchForQuery() {
+    List<Entity> expected = Arrays.asList();
+    List<Entity> actual = PostFilter.filterPosts("DNE");
```
Nit: make "DNE" a const? NO_QUERY_MATCH or something?

Commit:
Address Reviews:
+ Add Test Description to Test Class
+ Use 'Enter' code instead of key code
+ Remove the use of weights and use Optionals instead
+ Name change Gson to Json

Commit:
Use single quotes

lyndapham commented on capstone_project/src/test/java/com/google/sps/PostFilterTest.java:
```diff
@@ -102,7 +110,7 @@ public void returnsAllPostsInPriorityOrderWithQuery() {
   public void returnsEmptyArrayOnNoMatchForQuery() {
     List<Entity> expected = Arrays.asList();
-    List<Entity> actual = PostFilter.filterPosts("DNE");
+    List<Entity> actual = PostFilter.filterPosts(NON_EXISTANT_QUERY);
```
Nit: NON_EXISTENT... with an e

lyndapham commented:
lgtm!

Commit:
Fix typo

andrewnguyen commented on capstone_project/src/main/java/com/google/sps/data/PostFilter.java:
```diff
@@ -0,0 +1,74 @@
+    } else if (postText.contains(lowerQuery)) {
+      return Optional.of(-1);
+    }
+    return Optional.empty();
```
nice
